rest/admin_api.go

```diff
@@ -1423,78 +1424,123 @@ func (h *handler) handleDeleteDB() error {
+	)
 
-	if redact && includeDoc && includeDocSet {
-		return base.HTTPErrorf(http.StatusBadRequest, "redact and include_doc cannot be true at the same time. "+
-			"If you want to redact you must specify include_doc=false")
+	if redact {
+		// cannot explicitly include doc and redact - it doesn't make sense since the doc body itself is considered UserData and eligible for redaction.
+		if includeDocSet && includeDoc {
+			return base.HTTPErrorf(http.StatusBadRequest, "redact and include_doc cannot be true at the same time. "+
+				"If you want to redact you must specify include_doc=false")
+		}
+		// if include_doc was not explicitly set, and we want to redact, then automatically set include_doc to false
+		includeDoc = false
+	} else {
+		if redactSalt != "" {
+			return base.HTTPErrorf(http.StatusBadRequest, "redactSalt can only be used when redact=true")
+		}
 	}
 
-	if redact && !includeDocSet {
-		includeDoc = false
+	// collect the set of xattrs to fetch that Sync Gateway is interested in.
+	xattrKeys := []string{
+		base.SyncXattrName,
+		base.GlobalXattrName,
+		base.MouXattrName,
+		base.VvXattrName,
+	}
+	userXattrKey := h.db.Options.UserXattrKey
+	if userXattrKey != "" {
+		// we'll redact this later after fetching - it's still useful to know it was present even if we can't see the contents.
+		xattrKeys = append(xattrKeys, h.db.Options.UserXattrKey)
 	}
 
-	doc, err := h.collection.GetDocument(h.ctx(), docid, db.DocUnmarshalSync)
+	var (
+		err         error
+		docBody     []byte
+		xattrValues map[string][]byte
+	)
+	if includeDoc {
+		docBody, xattrValues, _, err = h.collection.GetCollectionDatastore().GetWithXattrs(h.ctx(), docID, xattrKeys)
+	} else {
+		xattrValues, _, err = h.collection.GetCollectionDatastore().GetXattrs(h.ctx(), docID, xattrKeys)
+	}
 	if err != nil {
 		return err
 	}
 
-	rawBytes := []byte(base.EmptyDocument)
-	if includeDoc {
-		if doc.IsDeleted() {
-			rawBytes = []byte(db.DeletedDocument)
-		} else {
-			docRawBodyBytes, err := doc.BodyBytes(h.ctx())
-			if err != nil {
-				return err
-			}
-			rawBytes = docRawBodyBytes
-		}
+	responseBody := []byte(base.EmptyDocument)
+	if docBody != nil {
+		responseBody = docBody
 	}
 
-	syncData := doc.SyncData
-	if redact {
-		if salt == "" {
-			salt = uuid.New().String()
+	xattrsObject := []byte(base.EmptyDocument)
+
+	// inject xattrs as top-level properties in the doc - this isn't *quite* representative of the actual stored doc,
+	// since they are outside the doc body as separate objects, but we'll stamp a marker value in each xattr object to tell us whether it was injected or inline in the original doc body.
+	for _, k := range xattrKeys {
+		v := xattrValues[k]
+		if !redact {
+			xattrsObject, err = base.InjectJSONProperties(xattrsObject, base.KVPair{Key: k, Val: json.RawMessage(v)})
+			if err != nil {
+				return base.HTTPErrorf(http.StatusInternalServerError, "couldn't inject xattr %q into response: %s", k, err)
+			}
+		} else {
```

Comment:
Will this write non redactable xattrs like mou if redact is true?

Reply:
Thanks!! This is a bug :)